### PR TITLE
Modify inference for predictions functionality

### PR DESCRIPTION
**Pull request by albertpod**
This PR aims at bringing the prediction functionality into `RxInfer.jl`.

### CODE REVIEW

File: src/inference.jl [hunk 1]
```diff
@@ -508,7 +524,13 @@ function inference(;
         returnvars   = Dict(variable => returnoption for (variable, value) in pairs(vardict) if (israndom(value) && !isanonymous(value)))
     end
 
+    # Check if `predictvars` is nothing but `data` has missing values
+    if predictvars === nothing
+        predictvars = Dict(variable => KeepLast() for (variable, value) in pairs(vardict) if (isdata(value) && !isempty(findall(ismissing, data[variable])) && !isanonymous(value)))
```
Comment (bvdmitri) [inline src/inference.jl:529]: Perhaps `any(ismissing, data[variable])` should be more efficient than `!isempty(findall(ismissing, data[variable]))`.

**Comment by albertpod** [comment]
Short summary of the modifications:

- The functions `make_actor` have been modified to accept `AbstractVariable` instead of `RandomVariable`.
- The `InferenceResult` struct has been expanded to include a new field named `predictions`.
- New functions `__inference_check_dataismissing`, `__inference_fill_predictions` have been added to handle predictions.
- The `inference` function has been updated to handle predictions. 
- The `data` argument is now optional if `predictvars` are specified.
- New logic has been added to handle prediction variables.
- Subscription logic has been updated to handle both types of variables.
- Several error checks and warnings have been added or modified to ensure that the data and prediction variables are correctly specified and handled.

File: src/inference.jl (hunk 1 above)
Comment (albertpod) [reply src/inference.jl]: It occurs in `__inference_check_dataismissing` now.

**Comment by albertpod** [comment]
**NOTE**:
The prediction functionality doesn't allow free energy computation at the moment.
Perhaps, we should just disable free energy or `ERROR` with the proper message that this feature isn't available now.
The research question remains, how do we compute FE on non-terminated graph (ping @ismailsenoz )

File: src/inference.jl [hunk 2]
```diff
@@ -141,6 +144,12 @@ function __inference_check_dicttype(keyword::Symbol, ::T) where {T}
           """)
 end
 
+__inference_check_dataismissing(d) = (ismissing(d) || any(ismissing, d))
+
+# Return NamedTuple for predictions
+__inference_fill_predictions(s::Symbol, d::AbstractArray) = NamedTuple{Tuple([s])}([repeat([missing], length(d))])
+__inference_fill_predictions(s::Symbol, d::DataVariable) = NamedTuple{Tuple([s])}([missing])
```
Comment (bvdmitri) [inline src/inference.jl:151]:
```suggestion
__inference_fill_predictions(s::Symbol, d::AbstractArray) = NamedTuple{(s, )}((Iterators.repeated(missing, length(d)), ))
__inference_fill_predictions(s::Symbol, d::DataVariable) = NamedTuple{(s, )}((missing, ))
```

for better perf and less memory allocs, but please double check that `Iterators.repeated` actually works

File: src/inference.jl [hunk 3]
```diff
@@ -512,7 +554,9 @@ function inference(;
     # catch exceptions during the inference procedure, optional, defaults to false
     catch_exception = false
 )
-    __inference_check_dicttype(:data, data)
+    if isnothing(data) && isnothing(predictvars)
+        error("""One of keyword arguments `data` or `predictvars`` must be specified""")
```
Comment (bvdmitri) [inline src/inference.jl:558]:
```suggestion
        error("""One of the keyword arguments `data` or `predictvars` must be specified""")
```

File: src/inference.jl [hunk 4]
```diff
@@ -492,6 +532,8 @@ function inference(;
     options = nothing,
     # Return structure info, optional, defaults to return everything at each iteration
     returnvars = nothing,
+    # Return structure info, optional, defaults to return everything at each iteration
```
Comment (bvdmitri) [inline src/inference.jl:535]:
```suggestion
    # Prediction structure info, optional, defaults to return everything at each iteration
```

File: src/inference.jl [hunk 5]
```diff
@@ -591,10 +686,15 @@ function inference(;
...
     try
         on_marginal_update = inference_get_callback(callbacks, :on_marginal_update)
-        subscriptions      = Dict(variable => subscribe!(obtain_marginal(vardict[variable]) |> ensure_update(fmodel, on_marginal_update, variable, updates[variable]), actor) for (variable, actor) in pairs(actors))
+        subscriptions_rv   = Dict(variable => subscribe!(obtain_marginal(vardict[variable]) |> ensure_update(fmodel, on_marginal_update, variable, updates[variable]), actor) for (variable, actor) in pairs(actors_rv))
+        subscriptions_pr   = Dict(variable => subscribe!(obtain_prediction(vardict[variable]) |> ensure_update(fmodel, on_marginal_update, variable, updates[variable]), actor) for (variable, actor) in pairs(actors_pr))
 
         is_free_energy, S, T = unwrap_free_energy_option(free_energy)
 
+        if !isempty(actors_pr) && is_free_energy
+            error("Cannot compute Bethe Free Energy for models with prediction variables. Please set `free_energy = false`.")
```
Comment (bvdmitri) [inline src/inference.jl:695]:
```suggestion
            error("The Bethe Free Energy computation is not compatible with the prediction functionality. Set `free_energy = false` to suppress this error.")
```

File: test/test_inference.jl [hunk 6]
```diff
@@ -720,4 +720,206 @@ end
...
+        θ ~ Beta(4.0, 8.0)
+        for i in 1:n
+            y[i] ~ Bernoulli(θ)
+        end
+    end
+
+    @test_throws ErrorException inference(model = coin_model(length(dataset)), data = (y = dataset,))
+
+    #test #10 free_energy error handling
+    @test_throws ErrorException inference(model = coin_model(length(dataset)), data = (y = dataset,), free_energy = true)
```
Comment (bvdmitri) [inline test/test_inference.jl:922]: Can we add one more tests for this case, with `data` which does not contain `missing`, but with explicit `predictvars`?

File: test/test_inference.jl [hunk 7]
```diff
@@ -894,6 +889,52 @@ end
...
+        end
+
+    end
+
+    result = inference(
+        model = coin_model(length(dataset)), 
+        data  = (y = dataset, )
+    )
+
+    @test typeof(last(result.predictions[:y])) <: Bernoulli
```
[the remark this answers is not in the document]
Comment (albertpod) [reply test/test_inference.jl:906]: added

File: test/test_inference.jl (hunk 6 above)
Comment (albertpod) [reply test/test_inference.jl:924]: added

**Pull request merged**